PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -204,80 +241,178 @@ public void AttachToRunspace(uint pid)
+                    // Initiate remote session with the remote machine
+                    _currentPowerShell.Runspace = _runspace;
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.EnterRemoteSessionDefaultCommand, remoteName));
+                    _currentPowerShell.Invoke();
+
+                    if (GetDebugScenario() == DebugScenario.Local)
+                    {
+                        return null;
+                    }
+
+                    // Run script on remote machine to grab all attachable processes
+                    _currentPowerShell.Commands.Clear();
+                    _currentPowerShell.AddScript(DebugEngineConstants.EnumerateRemoteProcessesScript);
+                    _currentPowerShell.Runspace = _runspace;
+                    Collection<PSObject> result = _currentPowerShell.Invoke();
+
+                    // Add each process' name and pid to the list to be returned
+                    for (int i = 0; i < result.Count; i += 2)
+                    {
+                        dynamic pid = result.ElementAt(i);
+                        dynamic name = result.ElementAt(i + 1);
+                        validProcesses.Add(new KeyValuePair<uint, string>(uint.Parse(pid.ToString()), name.ToString()));
+                    }
+
+                    // Exit the remote session and return results back to RemoteEnumDebugProcess
+                    _currentPowerShell.Commands.Clear();
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.ExitRemoteSessionDefaultCommand));
+                    _currentPowerShell.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    ServiceCommon.Log("Error connecting to remote machine; " + ex.ToString());
+                    return null;
+                }
+            }
+            return validProcesses;
         }
 
         /// <summary>
-        /// Detaches the HostService from a local runspace
+        /// Attaches the HostService to a remote runspace already in execution
         /// </summary>
-        public void DetachFromRunspace()
+        public void AttachToRemoteRunspace(uint pid, string remoteName)
         {
             if (_callback == null)
             {
                 _callback = OperationContext.Current.GetCallbackChannel<IDebugEngineCallback>();
             }
 
-            try
+            using (_currentPowerShell = PowerShell.Create())
             {
-                // attempt to gracefully detach the deugger
-                ClearBreakpoints();
-                ExecuteDebuggingCommand("detach", false);
+                try
+                {
+                    // enter into a remote session
+                    _currentPowerShell.Runspace = _runspace;
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.EnterRemoteSessionDefaultCommand, remoteName));
+                    _currentPowerShell.Invoke();

```

Comment:
As mentioned above, this would prompt for auth for the second time during the remote attaching process, which should be avoid.